src/token-processor/blockchain-api/blockchain-importer.ts

```diff
@@ -0,0 +1,109 @@
+import { ClarityAbi } from '@stacks/transactions';
+import { logger } from '../../logger';
+import {
+  BlockchainDbSmartContract,
+  PgBlockchainApiStore,
+} from '../../pg/blockchain-api/pg-blockchain-api-store';
+import { PgStore } from '../../pg/pg-store';
+import { isPgConnectionError } from '../../pg/postgres-tools/errors';
+import { timeout } from '../../pg/postgres-tools/helpers';
+import { waiter, Waiter } from '../util/helpers';
+import { getSmartContractSip } from '../util/sip-validation';
+
+export class SmartContractImportInterruptedError extends Error {
+  constructor() {
+    super();
+    this.name = this.constructor.name;
+  }
+}
+
+/**
+ * Imports token contracts and SIP-019 token metadata update notifications from the Stacks
+ * Blockchain API database.
+ */
+export class BlockchainImporter {
+  private readonly db: PgStore;
+  private readonly apiDb: PgBlockchainApiStore;
+  private importInterruptWaiter: Waiter<void>;
+  private importInterrupted = false;
+  private importFinished = false;
+
+  constructor(args: { db: PgStore; apiDb: PgBlockchainApiStore }) {
+    this.db = args.db;
+    this.apiDb = args.apiDb;
+    this.importInterruptWaiter = waiter();
+  }
+
+  async close() {
+    if (this.importFinished) {
+      return;
+    }
+    // Force the cursor to stop and wait.
+    this.importInterrupted = true;
+    await this.importInterruptWaiter;
+  }
+
+  async import() {
+    while (!this.importFinished) {
+      try {
+        const afterBlockHeight = (await this.db.getSmartContractsMaxBlockHeight()) ?? 1;
+        await this.importSmartContracts(afterBlockHeight);
+        this.importFinished = true;
+      } catch (error) {
+        if (isPgConnectionError(error)) {
+          logger.error(
+            `BlockchainImporter encountered a PG connection error during import, retrying...`,
+            error
+          );
+          await timeout(100);
+        } else if (error instanceof SmartContractImportInterruptedError) {
+          this.importInterruptWaiter.finish();
+          throw error;
+        } else {
+          throw error;
+        }
+      }
+    }
+  }
+
+  /**
+   * Scans the `smart_contracts` table in the Stacks Blockchain API postgres DB for every smart
+   * contract that exists in the blockchain. It then takes all of them which declare tokens and
+   * enqueues them for processing.
+   * @param afterBlockHeight - Minimum block height
+   */
+  private async importSmartContracts(afterBlockHeight: number) {
+    logger.info(
+      `BlockchainImporter smart contract import starting at block height ${afterBlockHeight}`
+    );
+    const cursor = this.apiDb.getSmartContractsCursor({ afterBlockHeight });
```

Comment:
What happens if the API ingests new blocks/contracts while this sql cursor is being iterated? Is the cursor smart enough to include those, or is there a check after this is completed to see if additional scan(s) need to be ran?

Reply:
Good catch. I made a change to enable the chain monitor while the import is taking place so it can catch any new contracts missed by the cursor while it's iterating.